tool_path_planner/src/edge_path_generator.cpp

```diff
@@ -0,0 +1,609 @@
+                                          const pcl::PointIndices& edge_indices,
+                                          std::vector<pcl::PointIndices>& segments)
+{
+  using namespace Eigen;
+  using namespace pcl;
+  using AlignedPointTVector = std::remove_reference<decltype(*this->octree_search_)>::type::AlignedPointTVector;
+  using PType = std::remove_reference<decltype(*this->input_cloud_)>::type::PointType;
+  struct Ray
+  {
+    PType p;
+    Vector3f d;
+  };
+
+  pcl::PointIndices current_segment_indices= edge_indices;
+  bool split = false;
+  int split_idx = -1;
+  for(std::size_t i =0 ; i < current_segment_indices.indices.size() -1; i++)
+  {
+    const PType& p1 = (*input_cloud_)[current_segment_indices.indices[i]];
+    const PType& p2 = (*input_cloud_)[current_segment_indices.indices[i+1]];
+
+    // check distance
+    Vector3f dir = (p2.getVector3fMap() - p1.getVector3fMap());
+    if(dir.norm() > config.octree_res * config.max_intersecting_voxels)
+    {
+      // splitting current segment into two
+      split = true;
+      split_idx = i;
+      CONSOLE_BRIDGE_logInform("Splitting at index %i, due to max dist exceeded %f",
+                               current_segment_indices.indices[split_idx],dir.norm());
+      break;
+    }
+
+    int voxels_encountered = checkSurfaceIntersection(config, p1,p2);
+    if(voxels_encountered > 0 && i > 0)
+    {
+      // splitting current segment into two
+      split = true;
+      split_idx = i;
+      CONSOLE_BRIDGE_logInform("Splitting at index %i, encountered %i voxels > %i ",
+                               current_segment_indices.indices[split_idx], voxels_encountered, config.max_intersecting_voxels);
+      break;
+    }
+  }
+
+  if(split)
+  {
+    pcl::PointIndices next_segment_indices;
+    auto next_segment_start_iter = std::next(current_segment_indices.indices.begin(),split_idx + 1);
+    next_segment_indices.indices.assign(next_segment_start_iter,current_segment_indices.indices.end());
+    current_segment_indices.indices.erase(next_segment_start_iter,current_segment_indices.indices.end());
+    segments.push_back(current_segment_indices);
+
+    if(next_segment_indices.indices.size() < 2)
+    {
+      CONSOLE_BRIDGE_logWarn("Only one point was found in the next segment, will ignore");
+      return true;
+    }
+
+    return splitEdgeSegments(config, next_segment_indices, segments);
+  }
+
+  segments.push_back(current_segment_indices);
+  return true;
+}
+
+int EdgePathGenerator::checkSurfaceIntersection(const tool_path_planner::EdgePathConfig& config,
+                                                const pcl::PointNormal& p1, const pcl::PointNormal& p2)
+{
+  using namespace Eigen;
+  using namespace pcl;
+  using PType = std::remove_const_t< std::remove_reference<decltype(p1)>::type >;
+  using AlignedPointTVector = std::remove_reference<decltype(*this->octree_search_)>::type::AlignedPointTVector;
+
+  struct Ray
+  {
+    PType p;
+    Vector3f d;
+  };
```

Comment:
Is there a reason you would not use a tuple instead of the struct?

Reply:
This struct makes it more convenient to access the members as opposed to a tuple where you'd have to use the `std::get<0>()` syntax